api/desecapi/serializers.py

```diff
@@ -184,13 +268,24 @@ def to_internal_value(self, data):
         ret = []
         errors = []
 
-        partial = self.partial
+        known_rrsets = [(x.subname, x.type) for x in self.instance] if isinstance(self.instance, list) else []
+        known_rrsets_index = {(x.subname, x.type): x for x in self.instance} if isinstance(self.instance, list) else {}
 
-        known_rrsets = {(x.subname, x.type) for x in self.instance} if isinstance(self.instance, list) else set()
+        partial = self.partial
 
-        for item in data:
+        for idx, item in enumerate(data):
             try:
-                self.partial = partial and (item.get('subname', ''), item.get('type', None)) in known_rrsets
+                key = (item.get('subname', ''), item.get('type', None))
+                self.partial = partial and key in known_rrsets
+                datas = {i: x for i, x in enumerate(data) if (x.get('subname', ''), x.get('type', None)) == key}
```

Comment:
I think the right-hand side is not used here, i.e. this dict can be replaced by `[i for i in enumerate(data) if ...]`. After making this change, another name would also be appropriate.

Reply:
Also, let's be conscious about the decision to introduce an algorithm with quadratic runtime here. (Just sayin', there will be requests which have a few thousand RRs, and we're paying quadratic price just for the convenience of listing the indexes of duplicates.)

One can do this in linear time time by creating an index with key subname/type and the value being a list of indices. The lookup time is then independent of the number of entries.

Reply:
code was outdated, commented on the new code